### ISSUE TITLE
nano-gpt 4b parametrs

### ISSUE THREAD

**Issue by yourourour-cmd** [codernunk/flick#2]
`# =========================================
# 4 BILLION PARAMETER TRANSFORMER (FROM ZERO)
# Single-file implementation
# =========================================

import math
import torch
import torch.nn as nn
import torch.nn.functional as F

# ========= CONFIG =========

VOCAB_SIZE = 32000
MAX_SEQ_LEN = 2048

D_MODEL = 4096        # hidden size
N_HEADS = 32
N_LAYERS = 20
FF_MULT = 4           # feedforward expansion

DROPOUT = 0.1
DEVICE = "cuda"
DTYPE = torch.float16

# ========= UTILS =========

class RMSNorm(nn.Module):
    def __init__(self, dim, eps=1e-6):
        super().__init__()
        self.weight = nn.Parameter(torch.ones(dim))
        self.eps = eps

    def forward(self, x):
        norm = x.pow(2).mean(-1, keepdim=True)
        return x * torch.rsqrt(norm + self.eps) * self.weight

# ========= ROTARY POSITION EMBEDDINGS =========

def rotary_embedding(dim, seq_len, device):
    inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2, device=device) / dim))
    t = torch.arange(seq_len, device=device)
    freqs = torch.einsum("i,j->ij", t, inv_freq)
    return torch.cos(freqs), torch.sin(freqs)

def apply_rotary(x, cos, sin):
    x1, x2 = x[..., ::2], x[..., 1::2]
    return torch.cat([x1 * cos - x2 * sin, x1 * sin + x2 * cos], dim=-1)

# ========= SELF ATTENTION =========

class SelfAttention(nn.Module):
    def __init__(self):
        super().__init__()
        self.head_dim = D_MODEL // N_HEADS

        self.qkv = nn.Linear(D_MODEL, D_MODEL * 3, bias=False)
        self.out = nn.Linear(D_MODEL, D_MODEL, bias=False)

    def forward(self, x, cos, sin):
        B, T, C = x.shape

        qkv = self.qkv(x)
        q, k, v = qkv.chunk(3, dim=-1)

        q = q.view(B, T, N_HEADS, self.head_dim).transpose(1, 2)
        k = k.view(B, T, N_HEADS, self.head_dim).transpose(1, 2)
        v = v.view(B, T, N_HEADS, self.head_dim).transpose(1, 2)

        q = apply_rotary(q, cos, sin)
        k = apply_rotary(k, cos, sin)

        attn = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(self.head_dim)
        mask = torch.tril(torch.ones(T, T, device=x.device)).bool()
        attn = attn.masked_fill(~mask, float("-inf"))

        attn = F.softmax(attn, dim=-1)
        out = torch.matmul(attn, v)

        out = out.transpose(1, 2).contiguous().view(B, T, C)
        return self.out(out)

# ========= FEEDFORWARD =========

class FeedForward(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(D_MODEL, D_MODEL * FF_MULT)
        self.fc2 = nn.Linear(D_MODEL * FF_MULT, D_MODEL)

    def forward(self, x):
        return self.fc2(F.gelu(self.fc1(x)))

# ========= TRANSFORMER BLOCK =========

class TransformerBlock(nn.Module):
    def __init__(self):
        super().__init__()
        self.norm1 = RMSNorm(D_MODEL)
        self.norm2 = RMSNorm(D_MODEL)
        self.attn = SelfAttention()
        self.ff = FeedForward()

    def forward(self, x, cos, sin):
        x = x + self.attn(self.norm1(x), cos, sin)
        x = x + self.ff(self.norm2(x))
        return x

# ========= GPT MODEL =========

class GPT4B(nn.Module):
    def __init__(self):
        super().__init__()
        self.embed = nn.Embedding(VOCAB_SIZE, D_MODEL)
        self.blocks = nn.ModuleList([TransformerBlock() for _ in range(N_LAYERS)])
        self.norm = RMSNorm(D_MODEL)
        self.lm_head = nn.Linear(D_MODEL, VOCAB_SIZE, bias=False)

        self.lm_head.weight = self.embed.weight  # weight tying

    def forward(self, tokens):
        B, T = tokens.shape
        x = self.embed(tokens)

        cos, sin = rotary_embedding(D_MODEL // N_HEADS, T, x.device)

        for block in self.blocks:
            x = block(x, cos, sin)

        x = self.norm(x)
        return self.lm_head(x)

# ========= TRAIN STEP =========

def train_step(model, tokens, optimizer):
    logits = model(tokens[:, :-1])
    loss = F.cross_entropy(
        logits.reshape(-1, VOCAB_SIZE),
        tokens[:, 1:].reshape(-1)
    )
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    return loss.item()

# ========= MAIN =========

if __name__ == "__main__":
    model = GPT4B().to(DEVICE).to(DTYPE)
    print(f"Model parameters: {sum(p.numel() for p in model.parameters()) / 1e9:.2f}B")

    # DUMMY INPUT (for test only)
    tokens = torch.randint(0, VOCAB_SIZE, (1, 128)).to(DEVICE)
    optimizer = torch.optim.AdamW(model.parameters(), lr=3e-4)

    loss = train_step(model, tokens, optimizer)
    print("Loss:", loss)
`

**Comment by codernunk** [comment]
Looks like spam.
